sklearn/decomposition/tests/test_nmf.py

```diff
@@ -93,74 +104,117 @@ def test_nls_close():
     A = np.abs(random_state.randn(30, 10))
-    nmf.ProjectedGradientNMF(n_components=15, sparseness='data',
-                             random_state=0).fit(A)
+    nmf.NMF(n_components=15, random_state=0, tol=1e-2).fit(A)
 
 
-def test_projgrad_nmf_sparseness():
+@ignore_warnings
+def test_nmf_sparseness():
     # Test sparseness
     # Test that sparsity constraints actually increase sparseness in the
     # part where they are applied.
+    tol = 1e-2
     A = np.abs(random_state.randn(10, 10))
-    m = nmf.ProjectedGradientNMF(n_components=5, random_state=0).fit(A)
+    m = nmf.ProjectedGradientNMF(n_components=5, random_state=0,
+                                 tol=tol).fit(A)
     data_sp = nmf.ProjectedGradientNMF(n_components=5, sparseness='data',
-                                       random_state=0).fit(A).data_sparseness_
+                                       random_state=0,
+                                       tol=tol).fit(A).data_sparseness_
     comp_sp = nmf.ProjectedGradientNMF(n_components=5, sparseness='components',
-                                       random_state=0).fit(A).comp_sparseness_
+                                       random_state=0,
+                                       tol=tol).fit(A).comp_sparseness_
     assert_greater(data_sp, m.data_sparseness_)
     assert_greater(comp_sp, m.comp_sparseness_)
 
 
+@ignore_warnings
 def test_sparse_input():
     # Test that sparse matrices are accepted as input
     from scipy.sparse import csc_matrix
 
     A = np.abs(random_state.randn(10, 10))
     A[:, 2 * np.arange(5)] = 0
-    T1 = nmf.ProjectedGradientNMF(n_components=5, init='random',
-                                  random_state=999).fit_transform(A)
-
     A_sparse = csc_matrix(A)
-    pg_nmf = nmf.ProjectedGradientNMF(n_components=5, init='random',
-                                      random_state=999)
-    T2 = pg_nmf.fit_transform(A_sparse)
-    assert_array_almost_equal(pg_nmf.reconstruction_err_,
-                              linalg.norm(A - np.dot(T2, pg_nmf.components_),
-                                          'fro'))
-    assert_array_almost_equal(T1, T2)
 
-    # same with sparseness
+    for solver in ('proj-grad', 'coordinate'):
+        est1 = nmf.NMF(solver=solver, n_components=5, init='random',
+                       random_state=0, tol=1e-2)
+        est2 = clone(est1)
 
-    T2 = nmf.ProjectedGradientNMF(
-        n_components=5, init='random', sparseness='data',
-        random_state=999).fit_transform(A_sparse)
-    T1 = nmf.ProjectedGradientNMF(
-        n_components=5, init='random', sparseness='data',
-        random_state=999).fit_transform(A)
+        W1 = est1.fit_transform(A)
+        W2 = est2.fit_transform(A_sparse)
+        H1 = est1.components_
+        H2 = est2.components_
 
+        assert_array_almost_equal(est2.reconstruction_err_,
+                                  nmf._safe_compute_error(A, W2, H2))
```

Comment:
This is only supposed to test `_safe_compute_error`, right? Could you move this into a separate test that does only that? The sparse frobenius norm could become an internal util at some point.